reasoner/planner/RecursivePlanner.java

```diff
@@ -0,0 +1,382 @@
+                    plan(callMode); // Acyclic dependencies can be fully planned
+                }
+            }
+        } else if (resolvable.isNegated()) {
+            iterate(resolvable.asNegated().disjunction().conjunctions()).forEachRemaining(conjunction -> {
+                CallMode callMode = new CallMode(conjunction, resolvableMode);
+                recursivelyGeneratePlanChoices(callMode);
+                plan(callMode);
+            });
+        }
+    }
+
+    private void recursivelyGeneratePlanChoices(CallMode callMode) {
+        if (!planChoices.containsKey(callMode)) {
+            planChoices.put(callMode, null); // Guard
+            ConjunctionNode conjunctionNode = conjunctionGraph.conjunctionNode(callMode.conjunction);
+            answerCountEstimator.buildConjunctionModel(callMode.conjunction);
+
+            Map<Set<Pair<Concludable, Set<Variable>>>, PlanChoice> planChoices = new HashMap<>();
+            PartialOrderReductionSearch porSearch = new PartialOrderReductionSearch(logicMgr.compile(callMode.conjunction), callMode.mode);
+            for (List<Resolvable<?>> ordering : porSearch.allOrderings()) {
+                initialiseOrderingDependencies(conjunctionNode, ordering, callMode.mode);
+                PlanChoice planChoice = summarisePlanChoice(conjunctionNode, ordering, callMode.mode);
+
+                // Two orderings for the same CallMode with identical cyclic-concludable modes are interchangeable in the subgraph-plan
+                //      -> We only need to keep the cheaper one.
+                if (!planChoices.containsKey(planChoice.cyclicConcludableModes) ||
+                        planChoices.get(planChoice.cyclicConcludableModes).acyclicCost > planChoice.acyclicCost) {
+                    planChoices.put(planChoice.cyclicConcludableModes, planChoice);
+                }
+            }
+            this.planChoices.put(callMode, new HashSet<>(planChoices.values()));
+        }
+    }
+
+    private PlanChoice summarisePlanChoice(ConjunctionNode conjunctionNode, List<Resolvable<?>> ordering, Set<Variable> mode) {
+        Set<Variable> bounds = new HashSet<>(mode);
+        Set<Variable> restrictedBounds = new HashSet<>(); // Restricted by preceding resolvables
+        double acyclicCost = 0L;
+
+        Map<Concludable, Double> scalingFactors = new HashMap<>();
+        Set<Pair<Concludable, Set<Variable>>> cyclicConcludableModes = new HashSet<>();
+        AnswerCountEstimator.IncrementalEstimator estimator = answerCountEstimator.createIncrementalEstimator(conjunctionNode.conjunction());
+        for (Resolvable<?> resolvable : ordering) {
+            Set<Variable> estimateableVars = estimateableVariables(resolvable);
+            Set<Variable> resolvableMode = Collections.intersection(estimateableVars, bounds);
+            Set<Variable> restrictedResolvableBounds = Collections.intersection(estimateableVars, restrictedBounds);
```

Comment:
`bounds` -> Mode right?

Reply:
Unsure. Because a mode is like a function-signature. This is the set of variables bound within the conjunction.

Reply:
`restrictedVariables`?

Reply:
Yes, that would work

Reply:
Now I have restrictedVars, boundVars and Mode to mean 3 different things, but I like it.
restricted -> The space is restricted by this conjunction
bound -> During execution, this variable will have a concept associated with it.
mode -> The set of variables which will be bound when the rule is invoked.